### DISCUSSION TITLE
Graphql Filtering with variables in React

### DISCUSSION THREAD

**Discussion by asdutoit** [strapi/strapi#7851] [Legacy Topics]
I'm looking for some guidance on the best way to implement filter variables with React and graphql. 

I have written a query with React:

```javascript
const ONTS = gql`
  query ONTS(
    $filter_services_status: Boolean
    $filter_services_product: String
    $filter_addressUnit_addressComplex_siteName: String
  ) {
    onts(
      where: {
        services: {
          Status: $filter_services_status
          product: { description: $filter_services_product }
        }
        address_unit: {
          address_complex: {
            site_name: $filter_addressUnit_addressComplex_siteName
          }
        }
      }
    ) {
      id
      serial_number
      make
      model
      images {
        url
      }
      address_unit {
        unit_no
        address_complex {
          site_name
        }
        resident {
          id
          firstName
          lastName
          email
          cell_number
        }
      }
      services {
        Status
        product {
          description
          type
        }
        vlan {
          isp {
            name
          }
        }
      }
    }
  }
`;
```

In the Component, the user can select the filters to apply.  
Problem is, if the user only wants to use one filter, the "getOnts" useLazyQuery wants to run the variables for all the filters as apposed to only the filters selected.  IE, I only want the selected filters to be passed to the graphql query.  

```javascript
const FILTERS = {
  filter_services_status: "All",    {/* <------ Won't work as the filter requires boolean */}
  filter_services_product: "All",
  filter_addressUnit_addressComplex_siteName: "All",
};

const Toolbar = ({ getOnts }) => {
  const [filter, setFilter] = React.useState(FILTERS);

const onFilter = (e) => {
    setFilter({ ...filter, [e.target.name]: e.target.value });
  };

useEffect(() => {
   getOnts({variables: filter})
}, [filter])
```

I know the problem is with my useState hook that I already explicitly provide the filter variables and as a result, they will get pushed to the query.   What I don't know is the correct way to implement the query to only push the selected filters, without having to write a heap of "if" statements. 

Thanks in advance for any help. 

(PS) Maybe the Strapi API allows the "where" filter to accept "undefined" values and will then disregard that filter altogether?  "undefined" obviously doesn't work, but maybe there is a similar solution?

**Comment by derrickmehaffy** [comment]
Can you please test making the entire `where: $myFilter` a variable and pass your entire filter in the single variable.

**Reply by asdutoit** [reply]
Hi @derrickmehaffy ,  thanks I did try it and it does work great, but I am not sure how to dynamically build the JSON filter within react.

```javascript
const ONTS = gql`
  query ONTS($filter: JSON) {
    onts(where: $filter) {
      id
      serial_number
      make
      model
      images {
        url
      }
      address_unit {
        unit_no
        address_complex {
          site_name
        }
        resident {
          id
          firstName
          lastName
          email
          cell_number
        }
      }
      services {
        Status
        product {
          description
          type
        }
        vlan {
          isp {
            name
          }
        }
      }
    }
  }
`;
```

Here is a sample of the website with filters top left:

![Screenshot 2020-09-11 at 16 03 37](https://user-images.githubusercontent.com/20240370/92934967-71114000-f448-11ea-8a1f-91673306b9d8.png)

I am sure a ton of people have built these dynamic JSON filters, I just need a sample and I think (here's hoping) I should be able to figure it out. 

For example, look at my current code below.  Its a mess:

```javascript
const onFilter = (e) => {
    setFilter({ ...filter, [e.target.name]: e.target.value });
    if (
      e.target.name === "filter_services_status" &&
      e.target.value !== "All"
    ) {
      FILTER_JSON.services.Status = e.target.value;
    } else if (
      e.target.name === "filter_addressUnit_addressComplex_siteName" &&
      e.target.value !== "All"
    ) {
      FILTER_JSON.address_unit.address_complex.site_name = e.target.value;
    }
    console.log(FILTER_JSON);
  };

  useEffect(() => {
    if (
      filter.filter_services_status === "All" &&
      filter.filter_addressUnit_addressComplex_siteName === "All"
    ) {
      getOnts();
    } else {
      getOnts({ variables: FILTER_JSON });
    }
  }, [filter]);
```

**Reply by derrickmehaffy** [reply]
I'm not super familiar with the data tables in react but you would want to construct the filter object then pass that off to GraphQL to make the request. I'm not sure why it matters where in react you pass the filters but it's a topic that has come up before.

**Reply by asdutoit** [reply]
Thanks, @derrickmehaffy, 
I will investigate a bit further or just built a separate function to manage the dynamic filter.
